src/test/java/com/google/firebase/database/utilities/UtilitiesTest.java

```diff
@@ -185,4 +192,20 @@ public void onComplete(DatabaseError error, DatabaseReference ref) {
     assertSame(listener, result.getSecond());
   }
 
+  @Test
+  public void testExtractParamsFromUrl() {
+    Map<String, String> params = Utilities.getQueryParamsMap("abc=213&qpf=2312&xyz=true&qpf=hi");
+    assertEquals("213", params.get("abc"));
+    assertEquals("2312,hi", params.get("qpf"));
+    assertEquals("true", params.get("xyz"));
+  }
+
+
+  @Test
+  public void testExtractParamsFromEncodedUrl() {
+    Map<String, String> params = Utilities.getQueryParamsMap("abc=213&qpf=2312&xyz=true&qpf=hi");
```

Comment:
Hey, isn't this the same as the one above?

Reply:
Also good stuff to test: `a=%3F%3F%3F&b=%3D%26%3D` should give you `???` for a and `=&=` for b. Also test this with the URI class and see if it does the right thing.

Reply:
As a extreme case, you should be able to put a full URL in a param if you encode it correctly :)

Reply:
Thanks for this @yuchenshi. I switched up the behavior: `Utilities.getQueryParamsMap` now accepts an encoded query string and decodes each key-value pair separately.